### ISSUE TITLE
Documentation: How are Pandas Categorical features identified, by name or the underlying categorical code?

### ISSUE THREAD

**Issue by AllanLRH** [microsoft/LightGBM#2761]
I can't find an answer to this in the documentation or on the internet.

How are entries in a categorical column identified in LightGBM when passing a Pandas DataFrame with columns encoded as categoricals?

I'm running a ML model based on LightGBM in a production environment, and my data contains categorical variables.
Insted af label-encoding the categorical variables and specifying a list of categorical varaibles (columns) to LightGBM, I'm using Pandas DataFrames with the categorical columns set to be of the categorical dtype (`df['some_categorical_column'] = df['some_categorical_column'].astype('category')` ).
LightGBM can figure out which columns are categorical by inspecting the DataFrame, but **does it assume that the [underlying categorical codes](http://pandas.pydata.org/pandas-docs/stable/user_guide/categorical.html) are the same as they were when the model were trained, or does it inspect the associated labels for each categorical column?**

Since my setup involves restarting the Python interpreter and reloading data, I'm concerned that the mapping of categorical values and their associated codes might change each time I run the code, and that might make LightGBM consider _Green_ to be _Red_, _Yellow_ to be _Blue_ and so on:

```
# Mapping for a single categorical column when the LightGBM-model is trained:
Red:     1
Blue:    2
Green:   3
Yellow:  4


# Possible mapping for a single categorical column when the LightGBM-model is running in production:
Green:   1
Blue:    2
Yellow:  3
Red:     4
```

## Environment info

Operating System:
Linux (Ubuntu)

CPU/GPU model:
AMD Threadripper

C++/Python/R version:
Python 3.7.3

LightGBM version or commit hash:
2.3.1

## Error message

N/A

## Reproducible examples

N/A

## Steps to reproduce

N/A

**Comment by StrikerRUS** [comment]
Please refer to the code:

https://github.com/microsoft/LightGBM/blob/704e08300ead3e70625d5431c845141f2e564123/python-package/lightgbm/basic.py#L505

https://github.com/microsoft/LightGBM/blob/704e08300ead3e70625d5431c845141f2e564123/python-package/lightgbm/basic.py#L324-L331

I believe that within the same version pandas categories mapping is deterministic.

**Comment by StrikerRUS** [comment]
> By default the categories are ordered alphabetically, which is why in the example below Amphibian is represented by a zero.
http://benalexkeen.com/mapping-categorical-data-in-pandas/

**Comment by AllanLRH** [comment]
All right, so the categorical data will be converted correctly by virtue of Pandas categorical codes to the data in a sorted fashion.

However, if previously unseen data are introduced during the deployment period of the model, it seems like we'd have to manage the categorical codes manually:

```
X_train LGB-transformed codes and original Series
(array([2.], dtype=float32), 'D')
(array([0.], dtype=float32), 'A')
(array([1.], dtype=float32), 'C')
(array([1.], dtype=float32), 'C')
(array([3.], dtype=float32), 'E')
(array([3.], dtype=float32), 'E')
(array([0.], dtype=float32), 'A')

X_deploy LGB-transformed codes and original Series
(array([1.], dtype=float32), 'B')
(array([4.], dtype=float32), 'E')
(array([0.], dtype=float32), 'A')
(array([2.], dtype=float32), 'C')
(array([3.], dtype=float32), 'D')
(array([4.], dtype=float32), 'E')
(array([5.], dtype=float32), 'F')
```

Or put in another way: LightGBM relies on Pandas handling the encoding of the categoricals, and does not consider the actual labels of the Categorical when Label-encoding the data.

Thank for the help, it was invaluable!

I hope this example code will help others with similar questions.

<details>
  <summary>Click to see the code which generated the output from above</summary>
  
    #!/usr/bin/env python
    # -*- coding: utf8 -*-

    import pandas as pd
    import lightgbm as lgb

    X_train = pd.Series(['D', 'A', 'C', 'C', 'E', 'E', 'A'],
                        name='X', dtype='category').to_frame()
    X_deploy = pd.Series(['B', 'E', 'A', 'C', 'D', 'E', 'F'],
                         name='X', dtype='category').to_frame()

    X_train_transformed = lgb.basic._data_from_pandas(X_train, None, None, None)
    X_deploy_transformed = lgb.basic._data_from_pandas(X_deploy, None, None, None)

    # print("X_train", X_train, sep=':\n', end='\n\n')
    # print("X_deploy", X_deploy, sep=':\n', end='\n\n')

    # print('X_train_transformed', X_train_transformed, sep=':\n', end='\n\n')
    # print('X_deploy_transformed', X_deploy_transformed, sep=':\n', end='\n\n')


    print('X_train LGB-transformed codes and original Series')
    print(*list(zip(X_train_transformed[0], X_train.X.to_list())), sep='\n', end='\n\n')
    print('X_deploy LGB-transformed codes and original Series')
    print(*list(zip(X_deploy_transformed[0], X_deploy.X.to_list())), sep='\n', end='\n\n')
</details>

**Comment by ggmblr** [comment]
Sorry to bump this discussion, but i have a few follow up questions.

I'm running into a similar problem. Let's say i have a pandas DataFrame with a categorical column 'Country'.  What happens if there are more entries in the training set compared to the deploy one.
Take for example the following case:

```
X_train = pd.Series(['Germany','France','Italy','Spain','Belgium'],
                    name='X', dtype='category').to_frame()
X_deploy = pd.Series(['Germany','France','Spain','Belgium'],
                     name='X', dtype='category').to_frame()

# Pandas mapping for each df:

dict( enumerate(X_train['X'].cat.categories ) )
# {0: 'Belgium', 
#  1: 'France', 
#  2: 'Germany', 
#  3: 'Italy', 
#  4: 'Spain'}

dict( enumerate(X_deploy['X'].cat.categories ) )
# {0: 'Belgium', 
#  1: 'France', 
#  2: 'Germany', 
#  3: 'Spain'}
```

Thus if i now train my model, if i save the model to .json and look into the decision it made I'd find something of the type:

```
"right_child": {
    "split_index": 22,
    "split_feature": 1,
    "split_gain": 0.06720539927482605,
    "threshold": "0||2||3",
    "decision_type": "==",
    "default_left": false,
    "missing_type": "None",
    "internal_value": -0.00810006,
    "internal_weight": 771,
    "internal_count": 771,
```

I see that it made a decision based on my country feature, specifically if Country is Belgium, Germany or Italy, but it displays the category codes. 
Thus, if i now apply this model on my deploy set, with a .predict(), since the code 3 now pertains to Spain instead of Italy, will it wrongly compute the result?

If so how i can i avoid this?

Basically its reversing the question OP posed, and thus talking in the general sense.

**Comment by StrikerRUS** [comment]
@AllanLRH Sorry, just noticed that your code snippet is not fully correct.
Here, the last argument is not `None`, it will be old categories from the training phase.
```
X_deploy_transformed = lgb.basic._data_from_pandas(X_deploy, None, None, None)
```
I think that simplified version of `_data_from_pandas()` for one column will look like the following.
```
import numpy as np
import pandas as pd

X_train = pd.Series(['D', 'A', 'C', 'C', 'E', 'E', 'A'],
                    name='X', dtype='category').to_frame()
X_deploy = pd.Series(['B', 'E', 'A', 'C', 'D', 'E', 'F'],
                     name='X', dtype='category').to_frame()

# original mapping in training set
print(*list(zip(X_train['X'].values, X_train['X'].cat.codes)), sep='\n')
# ('D', 2)
# ('A', 0)
# ('C', 1)
# ('C', 1)
# ('E', 3)
# ('E', 3)
# ('A', 0)

# simplified version of _data_from_pandas() for one column
X_train_transformed = X_train.apply(lambda x: x.cat.codes).replace({-1: np.nan}).values

pandas_categorical = list(X_train['X'].cat.categories)
X_deploy['X'] = X_deploy['X'].cat.set_categories(pandas_categorical)
X_deploy_transformed = X_deploy.apply(lambda x: x.cat.codes).replace({-1: np.nan}).values

# mapping in deploy set
print(*list(zip(X_deploy['X'].values, X_deploy['X'].cat.codes)), sep='\n')
# (nan, -1)
# ('E', 3)
# ('A', 0)
# ('C', 1)
# ('D', 2)
# ('E', 3)
# (nan, -1)

# what actually will be passed to LightGBM
print(X_train_transformed)
# [[2]
#  [0]
#  [1]
#  [1]
#  [3]
#  [3]
#  [0]]

print(X_deploy_transformed)
# [[nan]
#  [ 3.]
#  [ 0.]
#  [ 1.]
#  [ 2.]
#  [ 3.]
#  [nan]]
```
